README.md
# Container Nursery
<p align="center"><img src="https://user-images.githubusercontent.com/2771251/132473388-9f0ff8d9-7bbb-47e5-b45f-9634d86a0dff.png" height="250"></p>

<p align="center">
  <img alt="GitHub package.json version" src="https://img.shields.io/github/package-json/v/ItsEcholot/ContainerNursery">
  <img alt="Maintenance" src="https://img.shields.io/maintenance/yes/2023">
</p>

Written in Node.js, this application acts as a HTTP reverse proxy and stops Docker containers which haven't been accessed recently and starts them again when a new request comes in. ContainerNursery also makes sure there are no more active WebSocket connections before stopping the container.

To improve the user experience a loading page is presented, which automatically reloads when the containers webserver is ready.

The application listens on port `80` by default for HTTP traffic and uses the socket at `/var/run/docker.sock` to connect to the docker daemon.

You can find the changelog in the [CHANGELOG.md](CHANGELOG.md) file.

## Demo


https://user-images.githubusercontent.com/2771251/132314400-817971fd-b364-4c78-9fed-650138960530.mp4


## Installation
I ***heavily*** recommend using another reverse proxy in front of ContainerNursery (for HTTPS, caching, etc.) pointing to your configured listening port (default `80`).

I also recommend running this application in a Docker container. Pull the latest image using:

```docker pull ghcr.io/itsecholot/containernursery:latest```

More information about the available tags and versions can be found on the [GitHub packages page](https://github.com/ItsEcholot/ContainerNursery/pkgs/container/containernursery).

### Example

```bash
docker run \
  --name='ContainerNursery' \
  -v /var/run/docker.sock:/var/run/docker.sock \
  -v /mnt/ContainerNursery/config:/usr/src/app/config \
  ghcr.io/itsecholot/containernursery:latest
```

## Configuration
To configure the proxy, edit the `config.yml` file in the `config` directory. The configuration file is automatically reloaded by the application when changes are made.
If no `config.yml` file is found an empty one is automatically created on application start.

The following top-level properties can be configured:
Property | Meaning
---------|--------|
`proxyListeningPort` | The port ContainerNursery should listen on for new http connections. Defaults to `80`.

The virtual hosts the proxy should handle can be configured by adding an object to the `proxyHosts` key.

The following properties are required:

| Property         | Meaning                                                                                                                                                                                                                                                                                                                                                                                           |
|------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `domain`         | Array or string containing domain(s) to listen for (equals the `host` header)                                                                                                                                                                                                                                                                                                                     |
| `containerName`  | Array or string of which container(s) (by name or id) to start and stop. ContainerNursery can start and stop multiple containers for a single proxy host. The first container in the list (main container) is used to check if the application is ready and reload the loading page. Note however that CN doesn't manage the timing of how the containers are started (database before app etc.). |
| `proxyHost`      | Domain / IP of container (use custom Docker bridge networks for dynDNS using the name of the container)                                                                                                                                                                                                                                                                                           |
| `proxyPort`      | Port on which the containers webserver listens on                                                                                                                                                                                                                                                                                                                                                 |
| `timeoutSeconds` | Seconds after which the container should be stopped. The internal timeout gets reset to this configured value every time a new HTTP request is made, or when the timer runs out while a Websocket connection is still active.                                                                                                                                                                     |

The following properties are optional:

| Property                       | Meaning                                                                                                                                                                                                                                                                                                                                                                                                  |
|--------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `displayName`      | The name (of the service) that should be displayed on the waiting page                                                                                                            |
| `proxyUseHttps`                | Boolean indicating if the proxy should use HTTPS to connect to the container. Defaults to `false`. This should only be used if the container only accepts HTTPS requests. It provides no additional security.                                                                                                                                                                                            |
| `stopOnTimeoutIfCpuUsageBelow` | If set, prevents the container from stopping when reaching the configured timeout if the averaged CPU usage (percentage between 0 and 100*core count) of the **main** container (first in the list of container names) is above this value. This is great for containers that should remain running while their doing intensive work even when nobody is doing any http requests, for example handbrake. |
| `proxyUseCustomMethod` | Can be set to a HTTP method (`HEAD`,`GET`, ...) which should be used for the ready check. Some services only respond to certain HTTP methods correctly. |

### Example Configuration
```yaml
proxyListeningPort: 80
proxyHosts:
  - domain: handbrake.yourdomain.io
    containerName: handbrake
    displayName: Handbrake
    proxyHost: localhost
    proxyPort: 5800
    timeoutSeconds: 15
    stopOnTimeoutIfCpuUsageBelow: 50
    proxyUseCustomMethod: GET
  - domain:
      - wordpress.yourdomain.io
      - wordpress.otherdomain.io
    containerName:
      - wordpress
      - mariadb
    proxyHost: wordpress
    proxyPort: 3000
    proxyUseHttps: true
    timeoutSeconds: 1800
```

Now point your existing reverse proxy for the hosts you configured in the previous step to the ContainerNursery IP/Domain Name on port 80.

**Important:** If you use the (otherwise) excellent [NginxProxyManager](https://github.com/jc21/nginx-proxy-manager) ***disable*** the caching for proxy hosts that are routed through ContainerNursery. Because the built-in caching config sadly also caches 404s this will lead to NginxProxyManager caching error messages while your app container is starting up and then refusing to serve the real `.js/.css` file once the startup is complete.

### Example Configuration for [NginxProxyManager](https://github.com/jc21/nginx-proxy-manager)

![NginxProxyManager Config](https://user-images.githubusercontent.com/2771251/132512090-621926eb-70b5-4801-a477-70cc300ab2a1.jpeg)

### Environment Variables
Additionally to the configuration done in the `congif.yml` file, there are a few settings which can only be configured by using environment variables.

| Name           | Valid Values                        | Description                                                                                                                                                                                         |
|----------------|-------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `CN_LOG_JSON`  | `true` / `false`                    | If set to `true` all logging is done in a machine readable format (JSON). Defaults to `false`.                                                                                                      |
| `CN_LOG_LEVEL` | `debug` / `info` / `warn` / `error` | Sets the minimum log level. Log entries below this importance level won't be printed to the console. Defaults to `info`.                                                                            |
| `CN_PORT`      | `integer`                           | Sets the port ContainerNursery listens on for new http connections. The `proxyListeningPort` option in the `config.yml` file takes precedence if both are set. Defaults to `80` if no value is set. |


CHANGELOG.md
Changelog
=========
**13.12.2023 - v1.9.0** Added the ability to define the name which should be displayed on the waiting page. Contributed by @Samurai1201 in #55.

**22.10.2023 - v1.8.0** Added the ability to define which HTTP method should be used during the ready check. Contributed by @Howard3 in #53.

**30.08.2022 - v1.7.0** Added the ability to proxy traffic to containers using HTTPS, while ignoring their certificate. This is for services which only listen for SSL requests. This does **not** provide any security over normal http proxying.

**08.02.2022 - v1.6.0** Added the ability to add multiple domains that point to the same proxy host.

**29.11.2021 - v1.5.2** Fixed a bug where misformed docker event payload would crash ContainerNursery when trying to parse JSON. Thanks to Alfy1080 on the Unraid Forums for the Bug report.

**12.10.2021 - v1.5.1** Fixed a bug where the loading page wouldn't be displayed when a path other than `/` is requested. Thanks to @JamesDAdams on GitHub for the Bug report.

**29.09.2021 - v1.5.0** Added the ability to stop (and start) multiple containers per proxy host. This is useful if the application supports multiple containers. The first container in the list is the main container, which is used to check if the container is ready and reload the loading page. For usage information check the README.md file on GitHub.
 
**24.09.2021 - v1.4.2** Handle SIGTERM. The ContainerNursery container should now stop (and thus also restart) much quicker.
 
**23.09.2021 - v1.4.1** Fixed an issue where certain editors broke the live config reload functionality by introducing a small delay before reading the config file.
 
**23.09.2021 - v1.4.0** Added stopOnTimeoutIfCpuUsageBelow setting to proxyHosts which prevents ContainerNursery from stoping containers if they're still busy (using more CPU than the limit). For usage information check the README.md file on GitHub.

src/index.ts
import 'dotenv/config';
import { createServer } from 'http';
import { createProxyServer } from 'http-proxy';
import express from 'express';
import logger from './Logger';
import ConfigManager from './ConfigManager';
import ProxyHost from './ProxyHost';

// Disable TLS certificate verification for the targets of the proxy server
process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

const proxyHosts: Map<string, ProxyHost> = new Map();
// eslint-disable-next-line no-unused-vars
const configManager = new ConfigManager(proxyHosts);
const proxy = createProxyServer({
  xfwd: true,
  secure: false
});

let proxyListeningPort = configManager.getProxyListeningPort();
const placeholderServerListeningPort = 8080;
const placeholderServerListeningHost = '127.0.0.1';

// Remove secure tag from cookies
proxy.on('proxyRes', (proxyRes) => {
  const sc = proxyRes.headers['set-cookie'];
  if (Array.isArray(sc)) {
    // eslint-disable-next-line no-param-reassign
    proxyRes.headers['set-cookie'] = sc.map(str => {
      return str.split(';')
        .filter(v => v.trim().toLowerCase() !== 'secure')
        .join('; ');
    });
  }
});

const stripPortHostHeader = (host: string | undefined): string | undefined => {
  if (!host) return host;
  return host.replace(/(:\d+)/i, '');
};

proxy.on('error', (err, req, res) => {
  req.headers.host = stripPortHostHeader(req.headers.host);
  logger.debug({ host: req.headers.host, error: err }, 'Error in proxying request');
  if (!req.headers.host) {
    res.writeHead(400, { 'Content-Type': 'text/plain' });
    res.write('Error: Request header host wasn\t specified');
    res.end();
    return;
  }
  const proxyHost = proxyHosts.get(req.headers.host as string);
  if (res.writeHead) {
    res.writeHead(500, { 'Content-Type': 'text/plain' });
    res.write(`Error: Host is not reachable ${JSON.stringify(proxyHost?.getTarget())}`);
    res.end();
  }
  logger.warn({ host: req.headers.host, target: proxyHost?.getTarget() }, 'Host not reachable');
});

const proxyServer = createServer((req, res) => {
  req.headers.host = stripPortHostHeader(req.headers.host);

  if (!req.headers.host) {
    res.writeHead(400, { 'Content-Type': 'text/plain' });
    res.write('Error: Request header host wasn\t specified');
    res.end();
    return;
  }

  let proxyHost;
  const urlParts = req.url ? req.url.split('/') : [''];

  for (let i = 0; i < urlParts.length; i += 1) {
    proxyHost = proxyHosts.get(req.headers.host + urlParts.filter((_, j) => j <= i).join('/'));
    if (proxyHost) break;
  }

  if (!proxyHost) {
    res.writeHead(400, { 'Content-Type': 'text/plain' });
    res.write(`Error: Proxy configuration is missing for ${req.headers.host}`);
    res.end();
    logger.warn({ host: req.headers.host, url: req.url }, 'Proxy configuration missing');
    return;
  }

  proxyHost.newConnection();
  proxy.web(req, res, {
    target: proxyHost.getTarget(),
    headers: proxyHost.getHeaders()
  });
  logger.debug({ host: req.headers.host, target: proxyHost.getTarget(), headers: proxyHost.getHeaders() }, 'Proxied request');
});

proxyServer.on('upgrade', (req, socket, head) => {
  req.headers.host = stripPortHostHeader(req.headers.host);
  if (!req.headers.host) {
    logger.warn('Socket upgrade failed, request header host not specified');
    return;
  }

  let proxyHost;
  const urlParts = req.url ? req.url.split('/') : [''];

  for (let i = 0; i < urlParts.length; i += 1) {
    proxyHost = proxyHosts.get(req.headers.host + urlParts.filter((_, j) => j <= i).join('/'));
    if (proxyHost) break;
  }

  if (!proxyHost) {
    logger.warn({ host: req.headers.host, url: req.url }, 'Socket upgrade failed, proxy configuration missing');
    return;
  }

  proxyHost.newSocketConnection(socket);
  proxy.ws(req, socket, head, {
    target: proxyHost.getTarget(),
    headers: proxyHost.getHeaders()
  });
  logger.debug({
    host: req.headers.host, url: req.url, target: proxyHost.getTarget(), headers: proxyHost.getHeaders()
  }, 'Proxied Upgrade request');
});

proxyServer.listen(proxyListeningPort);
logger.info({ port: proxyListeningPort }, 'Proxy listening');

configManager.on('port-update', () => {
  proxyListeningPort = configManager.getProxyListeningPort();
  proxyServer.close(() => {
    proxyServer.listen(proxyListeningPort);
    logger.info({ port: proxyListeningPort }, 'Proxy listening');
  });
});

const placeholderServer = express();
placeholderServer.set('views', 'views');
placeholderServer.set('view engine', 'ejs');
placeholderServer.use((_, res, next) => {
  res.setHeader('x-powered-by', 'ContainerNursery');
  next();
});
placeholderServer.get('*', (req, res) => {
  res.render('placeholder', { containerName: req.headers['x-container-nursery-container-name'] });
});
placeholderServer.listen(placeholderServerListeningPort, placeholderServerListeningHost);
logger.info({ port: placeholderServerListeningPort, host: placeholderServerListeningHost }, 'Proxy placeholder server listening');

process.on('SIGTERM', () => {
  process.exit(0);
});


src/ProxyHost.ts
import { ProxyTarget } from 'http-proxy';
import internal, { EventEmitter } from 'stream';
import fetch from 'node-fetch';
import logger from './Logger';
import DockerManager from './DockerManager';

const dockerManager = new DockerManager();

export default class ProxyHost {
  private domain: string[];
  private containerName: string[];
  private displayName: string[];
  private proxyHost: string;
  private proxyPort: number;
  public proxyUseHttps = false;
  public proxyUseCustomMethod: string | undefined;
  private timeoutSeconds: number;
  public stopOnTimeoutIfCpuUsageBelow = Infinity;

  private activeSockets: Set<internal.Duplex> = new Set();
  private containerEventEmitter: EventEmitter | null = null;
  private connectionTimeoutId: NodeJS.Timeout | null = null;
  private containerRunning: boolean | undefined = undefined;
  private containerReadyChecking = false;
  private startingHost = false;
  private stoppingHost = false;

  private cpuAverage = 0;
  private cpuAverageCounter = 0;
  private lastContainerCPUUsage = 0;
  private lastSystemCPUUsage = 0;

  constructor(
    domain: string[],
    containerName: string[],
    displayName: string[],
    proxyHost: string,
    proxyPort: number,
    timeoutSeconds: number
  ) {
    logger.info({
      host: domain,
      container: containerName,
      proxy: {
        host: proxyHost,
        port: proxyPort,
        useHttps: this.proxyUseHttps
      },
      timeoutSeconds: timeoutSeconds
    }, 'Added proxy host');

    this.domain = domain;
    this.containerName = containerName;
    this.displayName = displayName;
    this.proxyHost = proxyHost;
    this.proxyPort = proxyPort;
    this.timeoutSeconds = timeoutSeconds;
    dockerManager.isContainerRunning(this.containerName[0]).then(async (res) => {
      if (res) this.resetConnectionTimeout();
      this.containerRunning = res;

      const otherContainers = this.containerName.slice(1);
      const otherContainerChecks: Promise<boolean>[] = [];
      otherContainers.forEach(otherContainerName => {
        otherContainerChecks.push(dockerManager.isContainerRunning(otherContainerName));
      });
      (await Promise.all(otherContainerChecks)).forEach((otherContainerRunning, i) => {
        if (this.containerRunning === otherContainerRunning) return;
        if (otherContainerRunning) {
          logger.debug({ mainContainer: otherContainers[i], container: otherContainers[i] }, 'Stopping other container because main container isn\'t running');
          dockerManager.stopContainer(otherContainers[i]);
        } else {
          logger.debug({ mainContainer: otherContainers[i], container: otherContainers[i] }, 'Starting other container because main container is running');
          dockerManager.startContainer(otherContainers[i]);
        }
      });

      logger.debug({ container: this.containerName, running: res }, 'Initial docker state check done');
    });

    dockerManager.getContainerEventEmitter(this.containerName).then(eventEmitter => {
      this.containerEventEmitter = eventEmitter;
      eventEmitter.on('update', data => {
        logger.debug({ container: this.containerName, data }, 'Received container event');
        if (data.status === 'stop') {
          this.stopHost();
        } else if (data.status === 'start') {
          this.startHost();
        }
      });
    });

    dockerManager.getContainerStatsEventEmitter(this.containerName[0]).then(eventEmitter => {
      eventEmitter.on('update', data => {
        if (!this.containerRunning || !data.cpu_stats.cpu_usage.percpu_usage) return;

        this.calculateCPUAverage(
          data.cpu_stats.cpu_usage.total_usage,
          data.cpu_stats.system_cpu_usage,
          data.cpu_stats.cpu_usage.percpu_usage.length
        );
      });
    });
  }

  private async stopHost(): Promise<void> {
    if (this.stoppingHost) return;
    this.stoppingHost = true;

    this.containerRunning = false;
    this.stopConnectionTimeout();

    await Promise.all(this.containerName.map(async (container) => {
      if (await dockerManager.isContainerRunning(container)) {
        logger.info({ container, cpuUsageAverage: container === this.containerName[0] ? this.cpuAverage : undefined }, 'Stopping container');
        await dockerManager.stopContainer(container);
        logger.debug({ container }, 'Stopping container complete');
      }
    }));

    this.stoppingHost = false;
  }

  private async startHost(): Promise<void> {
    if (this.startingHost) return;
    this.startingHost = true;

    await Promise.all(this.containerName.map(async (container) => {
      if (!(await dockerManager.isContainerRunning(container))) {
        logger.info({ container }, 'Starting container');
        await dockerManager.startContainer(container);
        logger.debug({ container }, 'Starting container complete');
      }
    }));

    this.checkContainerReady();
    this.startingHost = false;
  }

  private checkContainerReady() {
    if (this.containerReadyChecking) return;

    this.containerReadyChecking = true;
    const checkInterval = setInterval(() => {
      this.resetConnectionTimeout();

      fetch(`http${this.proxyUseHttps ? 's' : ''}://${this.proxyHost}:${this.proxyPort}`, {
        method: this.proxyUseCustomMethod ?? 'HEAD'
      }).then(res => {
        logger.debug({
          domain: this.domain,
          proxyHost: this.proxyHost,
          proxyPort: this.proxyPort,
          status: res.status,
          headers: res.headers
        }, 'Checked if target is ready');

        if (res.status === 200 || (res.status >= 300 && res.status <= 399)) {
          clearInterval(checkInterval);
          this.containerReadyChecking = false;
          this.containerRunning = true;

          logger.debug({
            domain: this.domain,
            proxyHost: this.proxyHost,
            proxyPort: this.proxyPort
          }, 'Target is ready');
        }
      }).catch(err => logger.debug({ error: err }, 'Container readiness check failed'));
    }, 250);
  }

  private startConnectionTimeout(): void {
    this.connectionTimeoutId = setTimeout(
      () => this.onConnectionTimeout(), this.timeoutSeconds * 1000
    );
  }

  private resetConnectionTimeout(): void {
    logger.debug({
      domain: this.domain,
      timeoutSeconds: this.timeoutSeconds
    }, 'Resetting connection timeout');
    this.stopConnectionTimeout();
    this.startConnectionTimeout();
    this.resetCPUAverage();
  }

  private onConnectionTimeout(): void {
    if (this.activeSockets.size > 0) {
      logger.debug({
        domain: this.domain,
        activeSocketCount: this.activeSockets.size
      }, 'Reached timeout but there are still active sockets');
      this.resetConnectionTimeout();
      return;
    }
    if (this.cpuAverage > this.stopOnTimeoutIfCpuUsageBelow) {
      logger.debug({
        domain: this.domain,
        container: this.containerName[0],
        cpuUsageAverage: this.cpuAverage
      }, 'Reached timeout but the container cpu usage is above the minimum configured');
      this.resetConnectionTimeout();
      return;
    }

    this.stopHost();
  }

  private calculateCPUAverage(cpuUsage: number, systemCPUUsage: number, cpuCount: number): void {
    let cpuPercentage = 0;
    const cpuDelta = cpuUsage - this.lastContainerCPUUsage;
    const systemDelta = systemCPUUsage - this.lastSystemCPUUsage;

    if (cpuDelta > 0 && systemDelta > 0) {
      cpuPercentage = (cpuDelta / systemDelta) * cpuCount * 100;

      // using exponential weighted moving average
      const factor = 30; // if 1, then average = current value
      this.cpuAverageCounter += 1;
      this.cpuAverage += (cpuPercentage - this.cpuAverage) / Math.min(this.cpuAverageCounter, factor);
    }

    this.lastContainerCPUUsage = cpuUsage;
    this.lastSystemCPUUsage = systemCPUUsage;
  }

  private resetCPUAverage(): void {
    this.cpuAverage = 0;
    this.cpuAverageCounter = 0;
  }

  public getHeaders(): { [header: string]: string; } {
    const nameToUse = this.displayName[0] ?? this.containerName[0];
    return {
      'x-container-nursery-container-name': nameToUse
    };
  }

  public getTarget(): ProxyTarget {
    return {
      protocol: this.proxyUseHttps && this.containerRunning ? 'https:' : 'http:',
      host: this.containerRunning ? this.proxyHost : '127.0.0.1',
      port: this.containerRunning ? this.proxyPort : 8080
    };
  }

  public newConnection(): void {
    if (!this.containerRunning) {
      this.startHost();
    } else {
      this.resetConnectionTimeout();
    }
  }

  public newSocketConnection(socket: internal.Duplex): void {
    if (!this.containerRunning) {
      this.startHost();
    } else {
      this.resetConnectionTimeout();
    }

    this.activeSockets.add(socket);
    socket.once('close', () => {
      this.resetConnectionTimeout();
      this.activeSockets.delete(socket);
    });
  }

  public stopConnectionTimeout(): void {
    if (this.connectionTimeoutId) {
      clearTimeout(this.connectionTimeoutId);
      this.connectionTimeoutId = null;
    }
  }

  public stopContainerEventEmitter(): void {
    if (!this.containerEventEmitter) return;
    this.containerEventEmitter.emit('stop-stream');
    this.containerEventEmitter.removeAllListeners();
  }
}


src/DockerManager.ts
import EventEmitter from 'events';
import Docker from 'dockerode';
import logger from './Logger';

export default class DockerManager {
  private docker: Docker;

  constructor() {
    this.docker = new Docker({ socketPath: '/var/run/docker.sock' });
  }

  private findContainerByName(name: string): Docker.Container {
    return this.docker.getContainer(name);
  }

  public async isContainerRunning(name: string): Promise<boolean> {
    return (await this.findContainerByName(name).inspect()).State.Running;
  }

  public async startContainer(name: string): Promise<void> {
    return this.findContainerByName(name).start();
  }

  public async stopContainer(name: string): Promise<void> {
    return this.findContainerByName(name).stop();
  }

  public async getContainerEventEmitter(names: string[]): Promise<EventEmitter> {
    const eventEmitter = new EventEmitter();
    const readableStream = await this.docker.getEvents({
      filters: {
        container: names
      }
    });

    readableStream.on('data', chunk => {
      try {
        eventEmitter.emit('update', JSON.parse(chunk.toString('utf-8')));
      } catch (err) {
        logger.error(err, 'JSON parsing of Docker Event failed');
      }
    });

    eventEmitter.on('stop-stream', () => {
      readableStream.removeAllListeners();
    });

    return eventEmitter;
  }

  public async getContainerStatsEventEmitter(name: string): Promise<EventEmitter> {
    const eventEmitter = new EventEmitter();
    const statsStream = await this.findContainerByName(name)
      .stats({ stream: true }) as unknown as NodeJS.ReadableStream;

    statsStream.on('data', chunk => {
      try {
        eventEmitter.emit('update', JSON.parse(chunk.toString('utf-8')));
      } catch (err) {
        logger.error(err, 'JSON parsing of Docker Event failed');
      }
    });

    eventEmitter.on('stop-stream', () => {
      statsStream.removeAllListeners();
    });

    return eventEmitter;
  }
}


src/ConfigManager.ts
import fs from 'fs';
import Chokidar from 'chokidar';
import YAML from 'yaml';
import logger from './Logger';
import ProxyHost from './ProxyHost';
import EventEmitter from 'events';

const defaultProxyListeningPort = 80;
const placeholderServerListeningPort = 8080;

type ProxyHostConfig = {
  domain: string | string[]
  containerName: string | string[]
  displayName: string | string[]
  proxyHost: string
  proxyPort: number
  proxyUseHttps: boolean
  proxyUseCustomMethod: string
  timeoutSeconds: number
  stopOnTimeoutIfCpuUsageBelow?: number
}
type ApplicationConfig = {
  proxyListeningPort: number,
  proxyHosts: ProxyHostConfig[]
}

export default class ConfigManager {
  private configFile = 'config/config.yml';
  private proxyHosts: Map<string, ProxyHost>;
  private proxyListeningPort: number | null;
  private eventEmitter: EventEmitter;

  constructor(proxyHosts: Map<string, ProxyHost>) {
    this.proxyHosts = proxyHosts;
    this.proxyListeningPort = null;
    this.eventEmitter = new EventEmitter();
    this.createIfNotExist();
    this.parseConfig();
    this.watch();
  }

  private static parsePort(p: number): number | null {
    if (Number.isInteger(p) && p >= 0 && p <= 49151) {
      return p;
    }

    logger.warn({ portString: p }, 'Parsing proxy listening port failed! Is a valid value used (0-49151)?');
    return null;
  }

  // eslint-disable-next-line no-unused-vars
  public on(event: string, cb: (...args: unknown[]) => void): void {
    this.eventEmitter.on(event, cb);
  }

  public getProxyListeningPort(): number {
    if (this.proxyListeningPort
      && this.proxyListeningPort !== placeholderServerListeningPort) {
      return this.proxyListeningPort;
    }
    if (this.proxyListeningPort === placeholderServerListeningPort) {
      logger.warn({ placeholderServerListeningPort, desiredProxyListeningPort: this.proxyListeningPort }, "Can't use the same port as the internal placeholder server uses");
    }

    if (process.env.CN_PORT) {
      this.proxyListeningPort = ConfigManager.parsePort(parseInt(process.env.CN_PORT, 10));
      if (this.proxyListeningPort
        && this.proxyListeningPort !== placeholderServerListeningPort) {
        return this.proxyListeningPort;
      }
    }

    logger.warn({ port: defaultProxyListeningPort }, 'Using default proxy listening port');
    this.proxyListeningPort = defaultProxyListeningPort;
    return this.proxyListeningPort;
  }

  private createIfNotExist(): void {
    if (!fs.existsSync(this.configFile)) {
      fs.closeSync(fs.openSync(this.configFile, 'w'));
      logger.error('config.yml is missing, empty config file was created');
    }
  }

  private watch(): void {
    Chokidar.watch(this.configFile).on('change', () => {
      logger.info('Config changed, reloading hosts');
      setTimeout(() => this.parseConfig(), 500);
    });
  }

  private parseConfig(): void {
    const fileContent = fs.readFileSync(this.configFile, 'utf-8');
    const config: ApplicationConfig = YAML.parse(fileContent);

    if (!config || !config.proxyHosts) {
      logger.error({ invalidProperty: 'proxyHosts' }, 'Config is invalid, missing property');
    } else {
      this.loadProxyHosts(config.proxyHosts);
      if (config.proxyListeningPort) {
        const prevPort = this.proxyListeningPort;
        this.proxyListeningPort = ConfigManager.parsePort(config.proxyListeningPort);
        if (prevPort !== null && prevPort !== this.proxyListeningPort) {
          this.eventEmitter.emit('port-update');
        }
      }
    }
  }

  private loadProxyHosts(proxyHosts: ProxyHostConfig[]): void {
    logger.info('(Re)loading hosts, clearing all existing hosts first');
    this.clearOldProxyHosts();
    proxyHosts.forEach(proxyHostConfig => {
      if (!ConfigManager.validateProxyHost(proxyHostConfig)) {
        logger.error({ proxyHost: proxyHostConfig }, 'Config contains invalid proxyHost object');
      } else {
        const proxyHost = new ProxyHost( // TODO
          proxyHostConfig.domain instanceof Array
            ? proxyHostConfig.domain
            : [proxyHostConfig.domain],
          proxyHostConfig.containerName instanceof Array
            ? proxyHostConfig.containerName
            : [proxyHostConfig.containerName],
          proxyHostConfig.displayName instanceof Array
            ? proxyHostConfig.displayName
            : [proxyHostConfig.displayName],
          proxyHostConfig.proxyHost,
          proxyHostConfig.proxyPort,
          proxyHostConfig.timeoutSeconds
        );

        if (proxyHostConfig.proxyUseHttps) {
          proxyHost.proxyUseHttps = proxyHostConfig.proxyUseHttps;
        }

        if (proxyHostConfig.proxyUseCustomMethod) {
          proxyHost.proxyUseCustomMethod = proxyHostConfig.proxyUseCustomMethod;
        }

        if (proxyHostConfig.stopOnTimeoutIfCpuUsageBelow) {
          proxyHost.stopOnTimeoutIfCpuUsageBelow = proxyHostConfig.stopOnTimeoutIfCpuUsageBelow as number;
        }

        if (proxyHostConfig.domain instanceof Array) {
          proxyHostConfig.domain.forEach(domain => {
            this.proxyHosts.set(
              domain,
              proxyHost
            );
          });
        } else {
          this.proxyHosts.set(
            proxyHostConfig.domain as string,
            proxyHost
          );
        }
      }
    });
  }

  private clearOldProxyHosts(): void {
    this.proxyHosts.forEach(proxyHost => {
      proxyHost.stopConnectionTimeout();
      proxyHost.stopContainerEventEmitter();
    });
    this.proxyHosts.clear();
  }

  private static validateProxyHost(proxyHostConfig: Record<string, unknown>): boolean {
    // TODO
    if (!proxyHostConfig.domain) return false;
    if (!proxyHostConfig.containerName) return false;
    if (!proxyHostConfig.proxyHost) return false;
    if (!proxyHostConfig.proxyPort) return false;
    if (!proxyHostConfig.timeoutSeconds) return false;

    return true;
  }
}
